src/lib/Imp.hs

```diff
@@ -269,178 +307,311 @@ toImpHof env (maybeDest, hof) = do
+        Nothing -> do
+          (varMap, usedNames) <- get
+          let name = genFresh (Name AbstractedPtrName "ptr" 0) usedNames
+          put ( varMap    <> M.insert (ty, ptr) name varMap
+              , usedNames <> (name @> ()))
+          return $ Var $ name :> BaseTy (PtrType ty)
+    _ -> return $ Con $ Lit val
+  valsAndNames <- gets $ M.toAscList . fst
+  let impBinders = [Bind (name :> PtrType ty ) | ((ty, _), name) <- valsAndNames]
+  let vals = map (uncurry PtrLit . fst) valsAndNames
+  return (impBinders, vals, block')
   where
-    localRef refVar refDest m = withRefScope $ putRef refVar refDest >> m
+    initState :: (M.Map (PtrType, Ptr ()) Name, Env ())
+    initState = mempty
+
+traverseLiterals :: forall m . Monad m => Block -> (LitVal -> m Atom) -> m Block
+traverseLiterals block f =
+  liftM fst $ flip runSubstEmbedT  mempty $ traverseBlock def block
+  where
+    def :: TraversalDef (SubstEmbedT m)
+    def = (traverseDecl def, traverseExpr def, traverseAtomLiterals)
+
+    traverseAtomLiterals :: Atom -> SubstEmbedT m Atom
+    traverseAtomLiterals atom = case atom of
+      Con (Lit x) -> lift $ lift $ f x
+      _ -> traverseAtom def atom
 
 -- === Destination type ===
---
--- How is a destination different from a regular table? The fundamental distinction is
--- that destinations can be _written to_, and so just like tables yield scalars when
--- fully indexed, destinations simply yield _memory addresses_!
---
--- Based on that observation, we represent destinations using regular Imp atoms, with the
--- restriction that all atoms that terminate blocks appearing in the dest have to represent
--- dests or arrays. This allows us to nicely separate the logical layout, which is
--- encoded in table lambdas, from the physical layout of the data, which is encoded in the
--- array offset calculations that are based on the logical indices (i.e. lambda binders).
---
--- Note that we use a newtype to help distinguish between regular atoms and dests, since
--- it should only be allowed to convert from a dest to a regular atom (by simply issuing a
--- load instruction for each fully offset array), but not in the other direction.
--- Also, Dests should also always be fully simplified (i.e. no beta-redexes are allowed,
--- even under lambdas).
 
-newtype Dest = Dest Atom deriving (Show)
+type Dest = Atom  -- has type `Ref a` for some a
 type WithDest a = (Maybe Dest, a)
 
-deriving via Atom instance HasVars Dest
-
-pattern TabValAbs :: LamExpr -> Atom
-pattern TabValAbs a <- Lam a@(Abs _ (TabArrow, _))
+data DestEnv = DestEnv
+       { allocationInfo :: AllocInfo
+       , enclosingIdxs :: IndexStructure
+       , destDepVars :: Env () }
+
+-- The Cat env carries names for the pointers needed, along with their types and
+-- blocks that compute allocation sizes needed
+type DestM = ReaderT DestEnv (CatT (Env (Type, Block)) Embed)
+
+makeDest :: AllocInfo -> Type -> Embed ([(Binder, Atom)], Dest)
```

Comment:
Can you please add a comment documenting the return value? It's not immediately clear what should the binders represent.